### PR TITLE
Some changes for nicer category urls

### PR DESCRIPTION
**Pull request by david-mclean**
As discussed these are the changes for nicer category urls.

1 route test if failing now and I don't think it was before, the error its failing with is: 

Expected: <string.Empty>
But was:  <>

If you don't want to accept the pull based on this I understand.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,6 @@
 [Oo]bj
 [Bb]in
+[Aa]pp_Data
```
Comment (mikehadlow) [inline .gitignore:3]: Latest version has this, why not pull and merge.

File: Database/016_Add_UrlName_to_Category.sql [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Alter table [SutekiShop].[dbo].[Category]
+Add UrlName varchar(255) null,
+PrestaId int null
```
Comment (mikehadlow) [inline Database/016_Add_UrlName_to_Category.sql]: What is PrestaId for?

File: Suteki.Shop/Suteki.Shop.CreateDb/App.config [hunk 3]
```diff
@@ -1,7 +1,7 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <configuration>
   <connectionStrings>
-    <add name="SutekiShopConnectionString" connectionString="Data Source=localhost;Initial Catalog=SutekiShop;Integrated Security=True" providerName="System.Data.SqlClient" />
+    <add name="SutekiShopConnectionString" connectionString="Data Source=DAVIDS-PC\SQLEXPRESS;Initial Catalog=SutekiShopFork;Integrated Security=True" providerName="System.Data.SqlClient" />
```
Comment (mikehadlow) [inline Suteki.Shop/Suteki.Shop.CreateDb/App.config]: Machine specific configuration

File: Suteki.Shop/Suteki.Shop/Controllers/CategoryController.cs [hunk 4]
```diff
@@ -57,6 +58,9 @@ public ActionResult New([EntityBind(Fetch = false)] Category category)
             Validator.Validate(ModelState, () =>
 			    image = httpFileService.GetUploadedImages(Request, ImageDefinition.CategoryImage).SingleOrDefault());
 
+            if (string.IsNullOrWhiteSpace(category.UrlName))
+                category.UrlName = GetUrlSafeCategoryName(category.Name);
+
             if (!ModelState.IsValid)
```
Comment (mikehadlow) [inline Suteki.Shop/Suteki.Shop/Controllers/CategoryController.cs:61]: For consistencies sake, have a look at the way UrlName is handled for products (https://github.com/mikehadlow/sutekishop/blob/master/Suteki.Shop/Suteki.Shop/Models/Product.cs). UrlName is { get; protected set; } and every 'set' of Name updates UrlName. There's also an extension method (https://github.com/mikehadlow/sutekishop/blob/master/Suteki.Shop/Suteki.Shop/Models/ModelHelpers/UrlNameHelperExtensions.cs) that does the conversion of name to urlName.

File: Suteki.Shop/Suteki.Shop/Controllers/ProductController.cs [hunk 5]
```diff
@@ -49,10 +49,31 @@ public ActionResult Index(int id)
...
+
+			var products = category.Products.InOrder();
+
+			if (!userService.CurrentUser.IsAdministrator)
+			{
+				products = products.Active();
+			}
+
+			return View("Index", ShopView.Data.WithProducts(products).WithCategory(category));
+		}
```
Comment (mikehadlow) [inline Suteki.Shop/Suteki.Shop/Controllers/ProductController.cs:73]: Indentation looks a bit weird here ;)

File: Suteki.Shop/Suteki.Shop/Models/Category.cs [hunk 6]
```diff
@@ -8,18 +8,24 @@
...
-    public class Category : IActivatable, IOrderable, INamedEntity
+    public class Category : IActivatable, IOrderable, INamedEntity, IUrlNamed
     {
         public virtual int Id { get; set; }
 
         [Required(ErrorMessage = "Name is required")]
         public virtual string Name { get; set; }
 
+        public virtual int PrestaId { get; set; }
+
```
Comment (mikehadlow) [inline Suteki.Shop/Suteki.Shop/Models/Category.cs:28]: PrestaId again???

File: Suteki.Shop/Suteki.Shop/Models/Category.cs [hunk 7]
```diff
@@ -8,18 +8,24 @@
...
         public virtual int Position { get; set; }
         public virtual bool IsActive { get; set; }
         public virtual Image Image { get; set; }
         public virtual Category Parent { get; set; }
 
+        
+        [RegularExpression(@"[\w]+", ErrorMessage = "Only alpha-numeric and underscore characters allowed")]
+        public virtual string UrlName { get; set; }
+
         IList<Category> categories = new List<Category>();
```
Comment (mikehadlow) [inline Suteki.Shop/Suteki.Shop/Models/Category.cs:36]: See note about CategoryController

File: Suteki.Shop/Suteki.Shop/Suteki.Shop.csproj [hunk 8]
```diff
@@ -818,11 +818,11 @@
...
       <FlavorProperties GUID="{349c5851-65df-11da-9384-00065b846f21}">
         <WebProjectProperties>
-          <UseIIS>False</UseIIS>
+          <UseIIS>True</UseIIS>
           <AutoAssignPort>True</AutoAssignPort>
           <DevelopmentServerPort>5401</DevelopmentServerPort>
           <DevelopmentServerVPath>/</DevelopmentServerVPath>
-          <IISUrl>http://localhost/SutekiShop</IISUrl>
+          <IISUrl>http://localhost/SutekiShopFork</IISUrl>
           <NTLMAuthentication>False</NTLMAuthentication>
```
Comment (mikehadlow) [inline Suteki.Shop/Suteki.Shop/Suteki.Shop.csproj:857]: Careful, this shouldn't leak into the master branch

File: Suteki.Shop/Suteki.Shop/ViewData/CategoryViewData.cs [hunk 9]
```diff
@@ -12,6 +12,7 @@ public class CategoryViewData : IActivatable, IOrderable
         public int Position { get; set; }
         public bool IsActive { get; set; }
         public int? ImageId { get; set; }
+        public string UrlName { get; set; }
```
Comment (mikehadlow) [inline Suteki.Shop/Suteki.Shop/ViewData/CategoryViewData.cs:15]: See note on CategoryController.

File: Suteki.Shop/Suteki.Shop/Views/Category/Edit.aspx [hunk 10]
```diff
@@ -11,6 +11,7 @@
 		<%= this.Hidden(x => x.Category.Position) %>
 
 		<%= this.TextBox(x => x.Category.Name).Label("Name") %>
+        <%= this.TextBox(x=>x.Category.UrlName).Label("UrlName") %>
 		<%= this.Select(x => x.Category.Parent.Id).Options(Model.Categories, x => x.Id, x => x.Name).Label("Parent Category") %>
```
Comment (mikehadlow) [inline Suteki.Shop/Suteki.Shop/Views/Category/Edit.aspx:14]: User shouldn't be setting the UrlName, it should be derived from the Name property.

File: Suteki.Shop/Suteki.Shop/Web.config [hunk 11]
```diff
@@ -18,7 +18,7 @@
     </sectionGroup>
   </configSections>
   <connectionStrings>
-    <add name="SutekiShopConnectionString" connectionString="Data Source=localhost;Initial Catalog=SutekiShop;Integrated Security=True" providerName="System.Data.SqlClient" />
+    <add name="SutekiShopConnectionString" connectionString="Data Source=DAVIDS-PC\SQLEXPRESS;Initial Catalog=SutekiShopFork;Integrated Security=True" providerName="System.Data.SqlClient" />
```
Comment (mikehadlow) [inline Suteki.Shop/Suteki.Shop/Web.config:21]: Machine specific configuration.

**Comment by mikehadlow** [comment]
Thanks for this David, you've done some excellent work getting up to speed with the code and doing these changes. I've made some comments on the diffs. The main change is that I'd like you to use the same pattern that the product model uses for UrlName. You must get all the unit tests working as well before I'll merge it. Sorry if I seem a little strict, but I'd really like to get this in, it's such a worthwhile addition to suteki shop. Many thanks!

File: Suteki.Shop/Suteki.Shop/Views/Category/Edit.aspx (hunk 10 above)
Comment (david-mclean) [inline Suteki.Shop/Suteki.Shop/Views/Category/Edit.aspx:14]: I disagree with that, user should have control over the urlname so they can
adjust it to have the keyword phrase they require.

But if you dont want that in the release thats fair enough.
On Apr 20, 2012 8:51 AM, "Mike Hadlow" <
reply@reply.github.com>
wrote:

**Comment by david-mclean** [comment]
No i understand. If i was you I would request the same!

I dont fully understand why the test fails at the moment but will have to
get my head around it.

I will make the changes over the weekend.

The prestaid is from my project (thought id caught that), i need it to map
categories and products from a suppliers xml. I will be sure to remove!

Thanks,
David
On Apr 20, 2012 8:55 AM, "Mike Hadlow" <
reply@reply.github.com>
wrote:

File: .gitignore (hunk 1 above)
Comment (david-mclean) [inline .gitignore:3]: Thought I had the latest version when i added that.

David
On Apr 20, 2012 8:37 AM, "Mike Hadlow" <
reply@reply.github.com>
wrote:

File: Suteki.Shop/Suteki.Shop/Views/Category/Edit.aspx (hunk 10 above)
Comment (mikehadlow) [inline Suteki.Shop/Suteki.Shop/Views/Category/Edit.aspx:14]: Hmm, I don't think my users would understand what it was for. I'd quite like suteki shop to be as simple to use as possible without having to understand SEO principles. It should do a good job for SEO without the user really having to think about it.

**Comment by mikehadlow** [comment]
Thanks David, I'm looking forward to seeing the new version.

File: Suteki.Shop/Suteki.Shop/Views/Category/Edit.aspx (hunk 10 above)
Comment (david-mclean) [inline Suteki.Shop/Suteki.Shop/Views/Category/Edit.aspx:14]: I can see that.

I think that more advanced SEO options should you want them should be available, but I think thats someting for the future.

With that said I will go with good system defaults and no user input for now.

**Comment by david-mclean** [comment]
Closing this pull request.

**Pull request closed without merging**